src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DocumentMapping/RemoteDocumentMappingService.cs

```diff
@@ -5,22 +5,82 @@
 using System.Composition;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
 using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
 using Microsoft.CodeAnalysis.Razor.DocumentMapping;
 using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
 using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.Remote.Razor.DocumentMapping;
 
 [Export(typeof(IDocumentMappingService)), Shared]
 [method: ImportingConstructor]
 internal sealed class RemoteDocumentMappingService(
     IFilePathService filePathService,
+    IProjectCollectionResolver projectCollectionResolver,
     ILoggerFactory loggerFactory)
     : AbstractDocumentMappingService(filePathService, loggerFactory.GetOrCreateLogger<RemoteDocumentMappingService>())
 {
-    protected override ValueTask<RazorCodeDocument?> TryGetCodeDocumentAsync(Uri razorDocumentUri, CancellationToken cancellationToken)
+    private readonly IProjectCollectionResolver _projectCollectionResolver = projectCollectionResolver;
+
+    public async Task<(Uri MappedDocumentUri, LinePositionSpan MappedRange)> MapToHostDocumentUriAndRangeAsync(
+        RemoteDocumentSnapshot originSnapshot,
+        Uri generatedDocumentUri,
+        LinePositionSpan generatedDocumentRange,
+        CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var razorDocumentUri = FilePathService.GetRazorDocumentUri(generatedDocumentUri);
+
+        // For Html we just map the Uri, the range will be the same
+        if (FilePathService.IsVirtualHtmlFile(generatedDocumentUri))
+        {
+            return (razorDocumentUri, generatedDocumentRange);
+        }
+
+        // We only map from C# files
+        if (!FilePathService.IsVirtualCSharpFile(generatedDocumentUri))
+        {
+            return (generatedDocumentUri, generatedDocumentRange);
+        }
+
+        var razorFilePath = razorDocumentUri.GetAbsoluteOrUNCPath();
```

Comment:
We should probably ban this API in this project, as this is applying Razor Uri conventions, but we're in the Roslyn world now. Should use `RazorUri.GetDocumentFilePathFromUri(razorDocumentUri)` instead.

Reply:
Looks like you worked this out in the next commit :)

The actual logic in Roslyn is https://github.com/dotnet/roslyn/blob/main/src/LanguageServer/Protocol/Extensions/ProtocolConversions.cs#L190

Should probably create an extension method on Uri in this project to make things more discoverable.

Reply:
Yeah, as soon I tested my local Go to Definition changes against this branch, I immediately ran into the problem that I was _hoping_ you wouldn't spot. 😁 

Thanks for pointing out the Roslyn code. I'll make this better.